common/src/main/java/com/wynntils/core/config/ConfigManager.java

```diff
@@ -0,0 +1,178 @@
+            reader.close();
+            if (!fileElement.isJsonObject()) return; // invalid config file
+
+            // find configurable's section in file
+            JsonObject fileObject = fileElement.getAsJsonObject();
+            if (!fileObject.has(configurable.getJsonName())) return; // section doesn't exist, nothing to load
+            JsonObject configObject = fileObject.getAsJsonObject(configurable.getJsonName());
+
+            // attempt to load each value from the object
+            for (String optionName : configObject.keySet()) {
+                tryUpdateOption(configurable, optionName, configObject.get(optionName));
+            }
+        } catch (IOException e) {
+            Reference.LOGGER.error("Failed to load config: " + fileName);
+            e.printStackTrace();
+        }
+    }
+
+    private static void saveConfig(ConfigCategoryHolder category) {
+        String fileName = categoryToFileName(category.getName());
+        try {
+            // get file, create if necessary
+            File configFile = new File(userConfigs, fileName);
+            if (!configFile.exists()) configFile.createNewFile();
+
+            // create json object, with entry for each configurable
+            JsonObject configJson = new JsonObject();
+            for (ConfigurableHolder configurable : category.getConfigurables()) {
+                JsonObject configurableJson = configurableAsJson(configurable);
+                configJson.add(configurable.getJsonName(), configurableJson);
+            }
+
+            // write json to file
+            OutputStreamWriter fileWriter =
+                    new OutputStreamWriter(new FileOutputStream(configFile), StandardCharsets.UTF_8);
+            gson.toJson(configJson, fileWriter);
+            fileWriter.close();
+        } catch (IOException e) {
+            Reference.LOGGER.error("Failed to save config: " + fileName);
+            e.printStackTrace();
+        }
+    }
+
+    public static void saveConfigs() {
+        configCategories.forEach(ConfigManager::saveConfig);
+    }
+
+    public static ConfigCategoryHolder getOrCreateCategory(String name) {
+        for (ConfigCategoryHolder category : configCategories) {
+            if (category.getName().equals(name)) return category;
+        }
+
+        // category doesn't exist yet, has to be created
+        ConfigCategoryHolder newCategory = new ConfigCategoryHolder(name);
+        configCategories.add(newCategory);
+        return newCategory;
+    }
+
+    private static void tryUpdateOption(ConfigurableHolder configurable, String optionName, JsonElement valueJson) {
+        for (ConfigOptionHolder option : configurable.getOptions()) {
+            if (!optionName.equals(option.getOptionJsonName())) continue;
+
+            Object value = gson.fromJson(valueJson, option.getType());
+            option.setValue(value);
+            return;
+        }
+        // no matching option
```

Comment:
Maybe add a log here? "Failed to find `ConfigOptionHolder` for `{optionName}`".

Reply:
yeah that'd be nice